src/java/org/apache/cassandra/index/sai/plan/VectorTopKProcessor.java

```diff
@@ -147,6 +201,64 @@ public <U extends Unfiltered, R extends BaseRowIterator<U>, P extends BasePartit
         return new InMemoryUnfilteredPartitionIterator(command, unfilteredByPartition);
     }
 
+    private <U extends Unfiltered> List<Triple<PartitionInfo, Row, Float>> processPart(BaseRowIterator<U> part,
+                                                    SortedMap<PartitionInfo, TreeSet<Unfiltered>> unfilteredByPartition,
+                                                    AbstractQueue<Triple<PartitionInfo, Row, Float>> topK,
+                                                    boolean isParallel)
+    {
+        // compute key and static row score once per partition
+        DecoratedKey key = part.partitionKey();
+        Row staticRow = part.staticRow();
+        final PartitionInfo partitionInfo = PartitionInfo.create(part);
+        final float keyAndStaticScore = getScoreForRow(key, staticRow);
+
+        List<Triple<PartitionInfo, Row, Float>> res = new LinkedList<>();
+        while (part.hasNext())
+        {
+            Unfiltered unfiltered = part.next();
+            // Always include tombstones for coordinator. It relies on ReadCommand#withMetricsRecording to throw
+            // TombstoneOverwhelmingException to prevent OOM.
+            if (!unfiltered.isRow())
+            {
+                if (isParallel)
```

Comment:
my preference would be to use a concurrent structure (make unfilteredByPartition a ConcurrentMap ?) and not have to branch here

Reply:
I actually started with that, in profiles it was too hot. 
for the same reason I kept topK as PriorityQueue instead of changing it to PriorityBlockingQueue.

Reply:
I think we can simplify this a ton without concurrent structures or synchronized.  Pushed a commit showing what I mean.  It compiles but I have not tested it.